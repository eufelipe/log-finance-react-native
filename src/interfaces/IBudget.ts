import {Category} from 'models';

export default interface IBudget {
  description?: string;
  category: Category;
  value: number;
}
