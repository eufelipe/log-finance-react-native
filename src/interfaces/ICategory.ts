export default interface ICategory {
  id: number;
  description: string;
  key: string;
}
