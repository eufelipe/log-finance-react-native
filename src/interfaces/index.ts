export {default as ICategory} from './ICategory';
export {default as IEntry} from './IEntry';
export {default as IBudget} from './IBudget';
