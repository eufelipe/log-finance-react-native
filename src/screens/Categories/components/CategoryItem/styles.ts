import styled from 'styled-components/native';
import {TouchableOpacity, Text} from 'styles/layout';

export const Container = styled(TouchableOpacity)`
  margin: ${({theme}) => theme.spacing.great}px
    ${({theme}) => theme.spacing.small}px;
  align-items: center;
  flex-grow: 1;
  flex-basis: 0;
`;

export const Title = styled(Text)`
  font-size: ${({theme}) => theme.fontSizes.small};
  color: ${({theme}) => theme.colors.white};
  padding-top: ${({theme}) => theme.spacing.small}px;
`;
