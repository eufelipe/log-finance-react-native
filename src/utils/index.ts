export {default as number} from './number';
export {default as strings} from './strings';
