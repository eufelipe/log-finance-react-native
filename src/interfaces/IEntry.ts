import ICategory from './ICategory';

type EntryType = 'expense' | 'earning';

export default interface IEntry {
  id: number;
  description: string;
  type: EntryType;
  value: number;
  category: ICategory;
}
