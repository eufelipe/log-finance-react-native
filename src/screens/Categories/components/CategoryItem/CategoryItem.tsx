import React from 'react';
import {useTranslation} from 'react-i18next';
import {CategoryIcon} from 'components';

import {Container, Title} from './styles';
import {Category} from 'models';

interface Props {
  category: Category;
  onPressCategory: (category: Category) => void;
}

const CategoryItem = ({category, onPressCategory}: Props): JSX.Element => {
  const {t} = useTranslation('categories');

  const categoryKey = category.key;
  return (
    <Container
      onPress={() => onPressCategory(category)}
      testID="category-item-button"
      accessible={true}
      accessibilityLabel={`${t('accessibility-button-label')} ${t(
        categoryKey,
      )}`}
      accessibilityHint={`${t('accessibility-button-hint')} ${t(categoryKey)}`}
    >
      <CategoryIcon name={categoryKey} contained />
      <Title>{t(categoryKey)}</Title>
    </Container>
  );
};

export default CategoryItem;
