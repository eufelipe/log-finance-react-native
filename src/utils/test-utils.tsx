/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import {render, RenderAPI, RenderOptions} from '@testing-library/react-native';

import {theme} from 'styles/theme';
import {ThemeProvider} from 'styled-components/native';

import {NavigationContainer} from '@react-navigation/native';

interface Props {
  children: JSX.Element;
}

const AllTheProviders = ({children}: Props) => {
  return (
    <ThemeProvider theme={theme}>
      <NavigationContainer>{children}</NavigationContainer>
    </ThemeProvider>
  );
};

const customRender = (
  ui: React.ReactElement<any>,
  options?: RenderOptions,
): RenderAPI => render(ui, {wrapper: AllTheProviders, ...options});

// re-export everything
export * from '@testing-library/react-native';

// override render method
export {customRender as render};
