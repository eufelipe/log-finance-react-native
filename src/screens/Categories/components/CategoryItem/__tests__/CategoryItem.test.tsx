/* eslint-disable @typescript-eslint/unbound-method */
/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import {fireEvent} from '@testing-library/react-native';

import {render} from 'utils/test-utils';
import {ICategory} from 'interfaces';
import CategoryItem from '../CategoryItem';

jest.mock('react-native-vector-icons/MaterialIcons', () => 'Icon');

jest.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: (key: any) => {
      if (key === 'money') {
        return 'Restaurante';
      }
      return key;
    },
  }),
}));

const MOCK: ICategory = {
  id: 1,
  description: 'Restaurante',
  icon: 'money',
};

const onPressCategory = jest.fn();

describe('CategoryItem', () => {
  it('Test match snapshot CategoryItem', () => {
    // given
    const props = {
      category: MOCK,
      onPressCategory,
    };

    // when
    const {toJSON} = render(<CategoryItem {...props} />);

    // then
    expect(toJSON()).toMatchSnapshot();
  });

  it('Test if the description renders correctly', () => {
    // given
    const props = {
      category: MOCK,
      onPressCategory,
    };

    // when
    const {getByText} = render(<CategoryItem {...props} />);

    // then
    expect(getByText('Restaurante')).toBeTruthy();
    expect(() => getByText('no exist')).toThrow('Unable to find an element');
  });

  it('Tests if pressing the button is working correctly', () => {
    // given
    const props = {
      category: MOCK,
      onPressCategory,
    };

    // when
    const {getByTestId} = render(<CategoryItem {...props} />);
    const pressButton = getByTestId('category-item-button');
    fireEvent.press(pressButton);
    fireEvent.press(pressButton);

    // then
    expect(pressButton).toBeTruthy();
    expect(onPressCategory).toHaveBeenCalled();
    expect(onPressCategory).toHaveBeenCalledTimes(2);
  });
});
