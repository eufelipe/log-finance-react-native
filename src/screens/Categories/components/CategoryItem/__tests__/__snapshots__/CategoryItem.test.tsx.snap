// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryItem Test match snapshot CategoryItem 1`] = `
<View
  accessibilityHint="accessibility-button-hint Restaurante"
  accessibilityLabel="accessibility-button-label Restaurante"
  accessible={true}
  collapsable={false}
  focusable={true}
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "flexBasis": 0,
      "flexGrow": 1,
      "marginBottom": 21.428571428571427,
      "marginLeft": 10.714285714285714,
      "marginRight": 10.714285714285714,
      "marginTop": 21.428571428571427,
      "opacity": 1,
    }
  }
  testID="category-item-button"
>
  <View
    color="#6F7D9B"
    large={true}
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#6F7D9B",
          "borderRadius": 25,
          "height": 50,
          "justifyContent": "center",
          "width": 50,
        },
      ]
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "#ffffff",
            "fontSize": 24,
          },
          Array [
            Object {},
          ],
          Object {
            "fontFamily": "Material Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      î¢–
    </Text>
  </View>
  <Text
    style={
      Array [
        Object {
          "color": "#ffffff",
          "fontFamily": "Poppins-Regular",
          "fontSize": 28,
          "fontWeight": "400",
          "paddingTop": 10.714285714285714,
        },
      ]
    }
  >
    Restaurante
  </Text>
</View>
`;
